lib/executor/ramping_vus.go

```diff
@@ -636,40 +633,88 @@ func (vlv RampingVUs) Run(ctx context.Context, out chan<- stats.SampleContainer)
 		currentMaxAllowedVUs = newMaxAllowedVUs
 	}
 
-	handleAllRawSteps := func() bool {
-		for {
+	// This is the step that is the actual gracefulStop at the end
+	lastStep := gracefulExecutionSteps[len(gracefulExecutionSteps)-1]
+	// TODO maybe instead of getting only the last step (above) mergeSteps should return the
+	// remaining gracefulExecutionSteps? or maybe we should refactor this even more ...
+	mergedSteps := mergeSteps(rawExecutionSteps, gracefulExecutionSteps[:len(gracefulExecutionSteps)-1])
+	timer := time.NewTimer(time.Hour * 24)
+	for _, step := range mergedSteps {
+		offsetDiff := step.step.TimeOffset - time.Since(startTime)
+		if offsetDiff > 0 { // wait until time of event arrives // TODO have a mininum
+			timer.Reset(offsetDiff)
 			select {
-			case step, ok := <-rawStepEvents:
-				if !ok {
-					return true
-				}
-				handleNewScheduledVUs(step.PlannedVUs)
-			case step := <-gracefulLimitEvents:
-				if step.PlannedVUs > currentMaxAllowedVUs {
-					// Handle the case where a value is read from the
-					// gracefulLimitEvents channel before rawStepEvents
-					handleNewScheduledVUs(step.PlannedVUs)
-				}
-				handleNewMaxAllowedVUs(step.PlannedVUs)
-			case <-ctx.Done():
-				return false
+			case <-maxDurationCtx.Done():
+				return // exit if context is cancelled
+			case <-timer.C:
+				// now we do a step
 			}
 		}
+		if step.raw {
+			handleNewScheduledVUs(step.step.PlannedVUs)
+		} else {
+			handleNewMaxAllowedVUs(step.step.PlannedVUs)
+		}
 	}
 
-	if handleAllRawSteps() {
-		// Handle any remaining graceful stops
-		go func() {
-			for {
-				select {
-				case step := <-gracefulLimitEvents:
-					handleNewMaxAllowedVUs(step.PlannedVUs)
-				case <-maxDurationCtx.Done():
-					return
-				}
+	go func() {
+		offsetDiff := lastStep.TimeOffset - time.Since(startTime)
+		if offsetDiff > 0 { // wait until time of event arrives // TODO have a mininum
```

Comment:
Looking at this code, it seems almost identical to the one above, can you encapsulate it in a small `handleStep()` lambda? :sweat_smile: I know you aren't a big fan of lambdas, but they're still better than copy-pasting :wink:

Reply:
given that #1499 will probably be done soon after, and my expectation that most of Run will be completely rewritten with it, adding another lambda seems totally unnecessary for 4 lines of code

Reply:
fixed in 51b969a